librepo/handle.c

```diff
@@ -264,6 +266,32 @@ lr_handle_remote_sources_changed(LrHandle *handle, LrChangedRemoteSource type)
     }
 }
 
+struct callback_data {
+   GMainLoop *loop; 
+   long input_seconds;
+   time_t begin;
+   GNetworkMonitor *monitor;
+   GSocketConnectable *connectable; 
+   GCancellable *cancellable;
+};
+
+gboolean timeout_callback(gpointer data)
+{ 
+  struct callback_data *dt = (struct callback_data*)data;
+   if(!g_network_monitor_get_network_available(dt->monitor) &&
+      g_get_monotonic_time() - dt->begin <= dt->input_seconds * 1000000){
+      return TRUE;
+   } 
+   g_main_loop_quit(dt->loop);
```

Comment:
But wait didn't we lose the `g_network_monitor_can_reach()` bits?  I think we want to do e.g.

```
    if (g_network_monitor_can_reach (dt->monitor, dt->connectable, dt->cancellable, error))  { /* quit loop */ }
```